src/Npgsql/NpgsqlCommand.cs

```diff
@@ -1243,7 +1243,10 @@ internal async ValueTask<NpgsqlDataReader> ExecuteReader(CommandBehavior behavio
                     }
                     catch
                     {
-                        connector.CurrentReader = null;
+                        var reader = connector.CurrentReader;
+                        if (reader != null)
+                            await reader.Cleanup(async);
```

Comment:
It does call EndUserAction. But we also might fail before we have a reader.

Reply:
That is true. We can have a try/catch up to the point where the reader is created, and then move the code that cleans up the reader (if it exists) down to the common catch below - this would unify the multiplexing and non-multiplexing flow a bit (something we should improve further in the future).